Adds the grid check random event from baystation.

Shadowlight213 opened the pull request:
Ported from baystation12.
All apcs will shut down for between 30 to 60 seconds (15 to 30 for the smes)
The interfaces will be replaced with a failure notification and a button to manually reboot and end the failure early.
Machines with the critical_machine var set to true will cause their area to be exempted from the failure.
Currently tcomms machines and the PA (only if it is assembled) have these vars set.
The AI will also have their area be exempted.

:cl:
add: A new event has been added. The automated grid check. All apcs in non critical areas will be shut off for some time, but can be manually rebooted via interaction.
/:cl:

Shadowlight213 committed:
Adds the grid check random event from baystation.

Shadowlight213 committed:
Let's err on the side of caution and make the engineering smes immune.

optimumtact commented on code/modules/events/grid_check.dm:
```diff
@@ -0,0 +1,37 @@
+
+
+/datum/round_event/grid_check/start()
+	for(var/obj/machinery/power/smes/S in machines)
```
Make a global list of SMES and track/untrack them in new/destroy, the machine list can get quite long

Shadowlight213 replied:
K

RemieRichards commented on code/modules/events/grid_check.dm:
```diff
@@ -0,0 +1,37 @@
+
+			var/skip = 0
+			for(var/area_type in skipped_areas)
+				if(istype(A,area_type))
```
typecache.

Shadowlight213 replied:
What's a typecache?

RemieRichards replied:
you don't really need to know for this case, but you should learn.
swap 
```var/list/skipped_areas = list(/area/engine/engineering, /area/turret_protected/ai)``` 
for:
```var/static/list/skipped_areas = typecacheof(list(/area/engine/engineering, /area/turret_protected/ai))``` (yes ```/static``` works in procs)

and then swap
```if(istype(A, area_type))```
for:
```if(is_type_in_typecache(A, skipped_areas))```

Shadowlight213 replied:
Yeah I looked it up and figured it out. That's neat. get's rid of the for loop.

RemieRichards commented on code/modules/power/apc.dm:
```diff
@@ -626,13 +628,17 @@
 /obj/machinery/power/apc/ui_interact(mob/user, ui_key = "main", datum/tgui/ui = null, force_open = 0, \
 										datum/tgui/master_ui = null, datum/ui_state/state = default_state)
 	ui = SStgui.try_update_ui(user, src, ui_key, ui, force_open)
+	if(ui)
```
do this bit after the ```if(!ui)```, that way you cover both cases in 2 lines instead of 3

RemieRichards commented on code/modules/power/smes.dm:
```diff
@@ -341,14 +345,19 @@
 										datum/tgui/master_ui = null, datum/ui_state/state = default_state)
 	ui = SStgui.try_update_ui(user, src, ui_key, ui, force_open)
+	if(ui)
+		ui.set_autoupdate(state = (failure_timer ? 1 : 0))
```
same as previous comment

Shadowlight213 committed:
Improvements.

RemieRichards commented on code/modules/power/apc.dm:
```diff
@@ -628,12 +628,13 @@
 		ui = new(user, src, ui_key, "apc", name, 535, 515, master_ui, state)
 		ui.open()
 		ui.set_autoupdate(state = (failure_timer ? 1 : 0))
+	else
```
this was totally not my point.

see how you've reduced it from 3 lines to... 3 lines, as opposed to 2 how I said?
take the entire:
```
if(ui)
    ui.set_autoupdate(state = (failure_timer ? 1 : 0))
```
and put it down here, now it covers BOTH if the ui does and doesn't exist's cases, 2 lines, not 3.

Shadowlight213 replied:
Ohhhh

Shadowlight213 committed:
Oh.

Jalleo commented:
Uh dont we already have this event but its just a admin button at the moment?

Shadowlight213 commented:
@Jalleo This is slightly different from the admin button.
The admin version drains all apcs and smes completely and lasts forever until the power up button is pressed.
This version is temporary, doesn't drain the apcs or smes and each apc can be rebooted and reactivated early by any crewmember.

Shadowlight213 committed:
Removes some limits and ups the max disable time to 2 minutes.

Jalleo commented:
Probably could merge the two together but make the draining of the units a admin decision to activate when its activated by an admin.

Jazaen commented:
Hey, could you make it so that it checks for AIs in APC location (or APCs in location where AIs are), instead of just working on AI core? This could cuck robotics-made AIs, those that had to move due to tator attack or nukeops and those on maps without that area.

AnturK commented:
I'm not sure about this. Does this actually affect round in any way ? Seems like no-event in most scenarios.

lzimann commented:
What is even the point of this event? All I can see is being a major annoyance fnr

Kromgar commented:
All those dead people in cryo

Shadowlight213 commented:
@Jazaen
Those AIs should have enough time for the emergency thing to allow then to reboot the apc I think.

@lzimann 
complete darkness everywhere, a crowbar being all access, and stuff suddenly being powered off. Its like a mini powersink. Trust me, it created some chaos when I tested it.

@XDTM @kromgar hmm. That is a valid concern. If it is a major issue, I can try seeing if I can make a special exception for medbay and make it so equipment power stays on. Although I suppose that might be too lenient, so i'm 50/50 on that.

MrStonedOne commented:
The best way to do critical area exceptions is to add the exception to the device, not the apc. So that having tcoms or an ai or a cloning device attached the apc exempts the apc, rather than try to exempt the apc or the area.

Exempting smes should be done by detecting that the smes has no attached input(lore: its not attached to the grid). This covers gravgen, ai, tcoms, etc.

Exempting the sing is a bit harder, in fact, i think the best way would be to just not do this to smeses, as they either power apcs, or emitters, and if apcs are going off, then its all good.

Bawhoppen commented:
sounds fun

Shadowlight213 committed:
Removes SMES power failure.
changes exception system to use critical_machine variable on machines.
Tcomms machines and the AI have their areas exempted.
If the PA is assembled, engineering is exempted.

Shadowlight213 commented:
Ok. I removed the SMES power failure and made the changes mso suggested.

optimumtact commented on code/modules/events/grid_check.dm:
```diff
@@ -0,0 +1,19 @@
+	max_occurrences = 3
+
+/datum/round_event/grid_check
+	announceWhen	= 1
```
should these be true/false or are they actual count values?

Shadowlight213 replied:
They are actual count values. It's the delay after the event first triggers that the announcement happens.
If it's set to zero, there's no announcement though.
